news/_posts/2020-02-04-the-new-prestashop-test-framework.md

```diff
@@ -0,0 +1,95 @@
+
+
+But more importantly, it’s more stable in headless (which was a killer feature for us !) since it was designed with headless in mind.
+
+
+### New framework and architecture
+
+As mentioned, we wanted to use this opportunity to create a more robust way to create and maintain tests and avoid the issues encountered with the Selenium implementation. Since Puppeteer is just a component, QA team will create a complete framework around it, and we also introduced the use of [Page Object Pattern](https://martinfowler.com/bliki/PageObject.html). This pattern ensures you separate the test logic from the page logic.
+
+On one hand, you create separate objects for every page you need to automate things on. Each object will then hold its own logic, from selectors to basic methods to use on the page (for example, fill a form, sort a table, choose an option from a dropdown, etc).
+
+On the other hand, you create your test logic by using these objects. The goal is to have really small test steps which then use the methods from the objects. This way, your tests don’t know what’s happening behind the scenes. They only call methods which act as an abstraction layer.
+
+This is a good illustration of [Separation of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns): when a page is updated, you only change the corresponding object in its separate file so that selectors and methods still work the same. When your test needs an update, you only change the logic in the steps.
+
+
+### Re-focus the test effort
+
+The third point was about the campaign itself. We wanted to redirect the campaign to be more effective, faster (the full Selenium campaign could last for up to 20 hours!) and easier to maintain. That would mean a shift in the focus of the campaign: instead of making a “do-it-all” campaign, we split up the effort in 3 campaigns: 
+
+- A sanity tests campaign
+- An integration tests campaign
+- An end-to-end tests campaign
+
+Each of these campaigns has its perks.
+
+The **sanity campaign** is here to make sure developers do not break some core functionalities of PrestaShop when introducing changes. To be run for every Pull Request, it checks all the critical features of the application: the frontend catalog, the cart, the checkout process; but also the backend with the product section and the order section. It is a fast campaign (about 8 minutes) and it must pass for the PR to be reviewed: a fail in this section means being back to the coding board for the incriminated developer, with all the shame it brings (not really, but we QA engineers like to think it does).
+
+The **integration campaign** is the big pièce de résistance: every feature of every page is tested one by one (with of course a few exceptions when it’s too complicated to automate). The goal here is to verify that all features (taken individually) still work as expected. That means testing tables, filters, and CRUD functionalities (Create, Read, Update, Delete).
```

Comment:
> The **integration campaign** is the big pièce de résistance

Yeah, we got it, you live in Lyon where *everything you do is eating French food* but I think it's not the right place to write that 😅

Reply:
I actually proposed in the google doc "main course", that would be the equivalent of "plat de résistance". Simon rejected it :-)

Reply:
This one is not too bad, we can keep it if you want.

Reply:
https://www.merriam-webster.com/dictionary/pi%C3%A8ce%20de%20r%C3%A9sistance

Anyway I'll remove it, I don't want to be controversial.